GT-934: Implement Okta Authorization

celikceviren opened the pull request:
[Jira](https://jira.cru.org/browse/GT-934)

wrandall22 commented on src/app/app-routing.module.ts:
```diff
@@ -5,6 +5,7 @@ import { ResourcesComponent } from './components/resources/resources.component';
 import { LanguagesComponent } from './components/languages/languages.component';
 import { AttachmentsComponent } from './components/attachments/attachments.component';
 import { AuthGuardService } from './service/auth-guard/auth-guard.service';
+import { OklaLoginComponent } from './components/okla-login/okla-login.component';
```
Everywhere it says `Okla` it should say `Okta`

celikceviren replied:
Fixed. Sorry for the typo.

wrandall22 commented on src/environments/environment.staging.ts:
```diff
@@ -1,4 +1,12 @@
   base_url: 'https://mobile-content-api-stage.cru.org/',
+  oidc_auth: {
+    client_id: '0oa1ll7gg0vVRLctz0h8',
+    client_secret: '738c0f49-b689-437a-9920-4f4f1fbfb7bb',
```
@frett won't we want a separate app inside of Okta for each stage and prod?

frett replied:
I actually don't think we will need a separate app. The application only provides authentication, the API itself enforces authorization.

frett commented on src/app/service/auth/user-auth-session.service.ts:
```diff
@@ -0,0 +1,112 @@
+
+            const tToken = this._oauthService.getIdToken();
+            const tTokenObservable: Observable<AuthToken> = from(
+              this._authService.createAuthToken(tToken),
```
you didn't update the auth service to send the Okta `id_token` in as a `okta_id_token` attribute here:
https://github.com/CruGlobal/mobile-content-admin/blob/feature/GT-934/src/app/service/auth/auth.service.ts#L30

frett replied:
Spencer is going to be updating the API to take the `access_token` instead of the `id_token`, that will probably adjust the parameter name to be something else in the process

celikceviren replied:
@frett 
I did not update the `createAuthToken` method but now the app will pass the `id_token` to `createAuthToken` method and I presume this method will try to get an authentication token from the API. 
If this is successful, it will store the token to local storage and the API methods will add this token to the header of the requests. I did not change anything regarding this flow.
The only change I've made is, the app now will use `id_token` to get API authorization token and I think this is ok.
Is there anything that I'm missing here?

frett replied:
the `createAuthToken` method does need a minor adjustment. when we were using the shared secret to authenticate it would send the following request:
`{"data": {"attributes": {"code":"${accessCode}"}}}`

currently when sending the `id_token` it needs to send the following request:
`{"data": {"attributes": {"okta_id_token":"${accessCode}"}}}`

But, that will probably change one more time once our API developer has a chance to update the API to take the `access_token` instead.

frett replied:
This is the first mobile-content-api client that is using Okta for authentication, so there are some interoperability issues that we are sorting out as we go :)

celikceviren replied:
@frett 
Ok, I see. I missed that the parameter name would be different for getting the authorization code with okta token. 
I updated that with the latest commit. I also have a couple of more updates for displaying errors correctly.
Btw, I think it is better to use access_token for this, rather than id_token.

frett replied:
yeah, it's what the API team ended up using on a different API

frett commented:
@celikceviren I updated your logic to use the Okta `access_token` and was able to successfully login using it.

A few things I noticed:
- After I login, if I reload the page I need to login again.
  - it would be nice if it used a session cookie to store either the okta `access_token`/login state or the mobile-content-api token.
  - This used to work when using the pre-shared secret to authenticate
- There's been a few times I've ended up on the screen with the login button, but been unable to click the login button. Here is the only way I've figured out how to reproduce this:
  - Open a new incognito tab
  - Go directly to the URL: http://localhost:8080/
  - It will show a login error, and the login button isn't working after closing the error

celikceviren commented:
@frett 
Can you add me to the okta app again so that I can test and fix these?

frett commented:
> @frett
> Can you add me to the okta app again so that I can test and fix these?

yeah, let me go check your permissions, I thought you were still on the app, but something might have broken when I was adjusting rights the other day

frett commented:
> > @frett
> > Can you add me to the okta app again so that I can test and fix these?
> 
> yeah, let me go check your permissions, I thought you were still on the app, but something might have broken when I was adjusting rights the other day

@celikceviren it's showing that you are assigned to the application in Okta

celikceviren commented:
@frett 
I fixed the things with page reload & login button click.
I also added a new page: My Account. My account page will show user's name & email and contains a logout button.

frett commented:
everything looks good to me, I'll have Ann take a look at it

reldredge71 commented:
This looks good to me also from a JavaScript perspective.
